### PR TITLE
Fixed surface material overrides are not applied to the new mesh after changing mesh in the editor

### LINKED ISSUE #64163: (4.0) Surface material overrides are not applied instantly anymore when switching meshes
### Godot version

4.0.alpha13

### System information

linux

### Issue description

https://user-images.githubusercontent.com/71937306/183679276-d3a59887-300b-470b-b8cd-c584a0a7f9e9.mp4

- Saving then reloading the scene works.
- This works as expected in 3.5 official: the new mesh uses the surface material override instantly.

### Steps to reproduce

As seen in the video:
 1. Create a MeshInstance with any mesh,
 2. add a surface material override to it,
 3. swap the mesh with a new one.

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by nongvantinh**
Fixed https://github.com/godotengine/godot/issues/64163

### CODE REVIEW

File: scene/3d/mesh_instance_3d.cpp [hunk 1]
```diff
@@ -117,6 +117,12 @@ void MeshInstance3D::set_mesh(const Ref<Mesh> &p_mesh) {
 		mesh->connect(CoreStringNames::get_singleton()->changed, callable_mp(this, &MeshInstance3D::_mesh_changed));
 		_mesh_changed();
 		set_base(mesh->get_rid());
+		int surface_count = mesh->get_surface_count();
+		for (int surface_index(0); surface_index < surface_count; ++surface_index) {
+			if (surface_override_materials[surface_index].is_valid()) {
+				RS::get_singleton()->instance_set_surface_override_material(get_instance(), surface_index, surface_override_materials[surface_index]->get_rid());
+			}
+		}
```
Comment (clayjohn) [inline scene/3d/mesh_instance_3d.cpp:125]: This should move to be inside of ``_mesh_changed()`` so it is called everytime the mesh changes. For example, if a new surface is added to the mesh, you want this to be called, so it needs to be inside ``_mesh_changed()``

**Review by clayjohn — CHANGES_REQUESTED** [review]

File: scene/3d/mesh_instance_3d.cpp [hunk 2]
```diff
@@ -380,6 +380,13 @@ void MeshInstance3D::_mesh_changed() {
 		}
 	}
 
+	int surface_count = mesh->get_surface_count();
+	for (int surface_index(0); surface_index < surface_count; ++surface_index) {
```
Comment (clayjohn) [inline scene/3d/mesh_instance_3d.cpp:384]: Just a style nitpick, we don't use the ``surface_index(0)``-style initialization in for loops we do the old fashioned ``int surface_index = 0`` for consistency, we should do the same here.

Comment (nongvantinh) [reply scene/3d/mesh_instance_3d.cpp]: I thought Juan said `master` can use C++14/17 now.

Comment (nongvantinh) [reply scene/3d/mesh_instance_3d.cpp]: I need to do some tests with these changes. It seems like the solution doesn't work anymore.

**Review by clayjohn — DISMISSED** [review]

**Comment by nongvantinh** [comment]
Before we can update the properties of this mesh, we must `set_base` to notify the `Rendering Server` to know about it first.
It works smoothly now.

**Review by clayjohn — APPROVED** [review]
Looks good now! Thanks

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**